fixes ClickHouse/ClickHouse#87082 - materialized views to use database where they were created as context

dk-github opened the pull request:
fixes ClickHouse/ClickHouse#87082 - materialized views to use database where they were created as context

### Changelog category (leave one):
- Bug Fix (user-visible misbehavior in an official stable release)

### Changelog entry (a [user-readable short description](https://github.com/ClickHouse/ClickHouse/blob/master/docs/changelog_entry_guidelines.md) of the changes that goes into CHANGELOG.md):
Materialized views now use the database where they were created as execution context, meaning that:
- it is possible to omit explicit database qualification on names referred in view's select query
- if no explicit database qualification is given, the same database where the materialized view was created is assumed

### Documentation entry for user-facing changes

- [ ] Documentation is written (mandatory for new features)


<!-- ch-version-info:start -->
### Version info
- Merged into: `26.2.1.564`
<!-- ch-version-info:end -->

dk-github committed:
fixes ClickHouse/ClickHouse#87082 - materialized views to use database where they were created as context

dk-github commented:
hi @pamarcos , thanks for having a look!

Re illustrating that it fixes the issue:
- any advice on how best to add automated test for this? the error happens asyncrhonously at server side in refresh task scheduled by the server on view creation or on restart, so just running some SQL statements without waiting for reresh task to finish in background won't cut it...  
- I did test manually though, as in running the repo I shared in the issue, and verifying that with the fix it works as expected - can share the evidence here in a comment

Re regression:
- my understanding was that, now that you have added "can be tested" label, the regression tests will run - and if any fail, I will look at them and see if this was due to my change
- if there is anything else to it that I need to do - please let me know

dk-github committed:
fixes ClickHouse/ClickHouse#87082 - materialized views to use database where they were created as context - additional changes to handle edge cases like bootstrapping and restoring backup, where we are checking DDL for the materialized view, but the database where it will be running is not online yet

dk-github committed:
Merge remote-tracking branch 'origin/master' into issue-87082-fix-materialized-views-name-context

dk-github committed:
Merge branch 'ClickHouse:master' into issue-87082-fix-materialized-views-name-context

dk-github committed:
fixes ClickHouse/ClickHouse#87082 - materialized views to use database where they were created as context - adding test verifying the fix

pamarcos committed:
Merge branch 'master' into issue-87082-fix-materialized-views-name-context

pamarcos commented on tests/queries/0_stateless/03779_materialized_view_database_context.sql:
```diff
@@ -0,0 +1,66 @@
+-- Tags: no-parallel
```
I understand you added this because you're hardcoding database names. But we could use `{CLICKHOUSE_DATABASE:Identifier}` and `{CLICKHOUSE_DATABASE_1:Identifier}` instead so that they're random database names, allowing to execute the test in parallel.

pamarcos reviewed (commented):
I merged master into your branch to ensure we're running the CI with the updated fixes for flaky tests.

Apart of the nitpick Copilot pointed out of using `///` instead of `//` for comments, there's two things I see here:
- What about `ALTER TABLE ... MODIFY QUERY`? e.g.
```sql
ALTER TABLE db1_03779.mv1 MODIFY QUERY SELECT id, data, p, source FROM v2(p='after_alter');
```
- I think the tests can be improved to test what I understand this is trying to fix.

dk-github committed:
Merge branch 'ClickHouse:master' into issue-87082-fix-materialized-views-name-context

dk-github committed:
fixes ClickHouse/ClickHouse#87082 - materialized views to use database where they were created as context - addressing review comments

dk-github committed:
fixes ClickHouse/ClickHouse#87082 - materialized views to use database where they were created as context - adding fix and test for ALTER TABLE ... MODIFY QUERY case, which is a separate code path

dk-github commented on tests/queries/0_stateless/03779_materialized_view_database_context.sql:
```diff
@@ -0,0 +1,66 @@
+-- Parameterized view in db1_03779
+CREATE VIEW v1 AS SELECT id, data, {p:String} as p FROM tbl1;
+
+-- REFRESH MV calling unqualified parameterized view (should use db1_03779 context)
```
re ALTER TABLE ... MODIFY QUERY - good catch, thanks!  I concentrated on my own use case and didn't look carefully enough to realise that this MODIFY QUERY is even a thing, and that it uses a separate code path which needs the same fix - added now, and extended the test as well

dk-github committed:
Merge branch 'ClickHouse:master' into issue-87082-fix-materialized-views-name-context

dk-github committed:
Merge branch 'ClickHouse:master' into issue-87082-fix-materialized-views-name-context

pamarcos commented:
Some tests failed: https://s3.amazonaws.com/clickhouse-test-reports/json.html?PR=88193&sha=33b6fdfdb95d104d4e1514715b11084a21b85e49&name_0=PR

```
FAIL	03779_materialized_view_database_context retry_failed[cidb](https://play.clickhouse.com/play?user=play&run=1#V0lUSAogICAgOTAgQVMgaW50ZXJ2YWxfZGF5cwpTRUxFQ1QKICAgIHRvU3RhcnRPZkRheShjaGVja19zdGFydF90aW1lKSBBUyBkYXksCiAgICBjb3VudCgpIEFTIGZhaWx1cmVzLAogICAgZ3JvdXBVbmlxQXJyYXkocHVsbF9yZXF1ZXN0X251bWJlcikgQVMgcHJzLAogICAgYW55KHJlcG9ydF91cmwpIEFTIHJlcG9ydF91cmwKRlJPTSBjaGVja3MKV0hFUkUgKG5vdygpIC0gdG9JbnRlcnZhbERheShpbnRlcnZhbF9kYXlzKSkgPD0gY2hlY2tfc3RhcnRfdGltZQogICAgQU5EIHRlc3RfbmFtZSA9ICcwMzc3OV9tYXRlcmlhbGl6ZWRfdmlld19kYXRhYmFzZV9jb250ZXh0JwogICAgLS0gQU5EIGNoZWNrX25hbWUgPSAnU3RhdGVsZXNzIHRlc3RzIChhbWRfYmluYXJ5LCBvbGQgYW5hbHl6ZXIsIHMzIHN0b3JhZ2UsIERhdGFiYXNlUmVwbGljYXRlZCwgcGFyYWxsZWwpJwogICAgQU5EIHRlc3Rfc3RhdHVzIElOICgnRkFJTCcsICdFUlJPUicpCiAgICBBTkQgKHB1bGxfcmVxdWVzdF9udW1iZXIgPSAwIE9SIGJhc2VfcmVmIElOICgnbWFzdGVyJykpCiAgICBBTkQgdGVzdF9jb250ZXh0X3JhdyBMSUtFICclaGF2aW5nIHN0ZGVycm9yJScKR1JPVVAgQlkgZGF5Ck9SREVSIEJZIGRheSBERVNDCg==)	17.39
2026-01-13 16:26:10 Reason: having stderror:  
2026-01-13 16:26:10 [7ea13ad1ee08] 2026.01.13 14:26:01.323643 [ 293443 ] {0ff889a8-6707-4ca5-a695-ef4f661f35a7} <Error> executeQuery: Code: 46. DB::Exception: Unknown table function v2:  or incorrect parameterized view. (UNKNOWN_FUNCTION) (version 26.1.1.468) (from [::1]:56538) (comment: 03779_materialized_view_database_context.sh-test_79y7c1ek) (query 1, line 1) (in query: ALTER TABLE test_79y7c1ek_1.mv1 MODIFY QUERY SELECT id, data, p FROM v2(p='modified_query_db1')), Stack trace (when copying this message, always include the lines below):
2026-01-13 16:26:10 
2026-01-13 16:26:10 0. ./contrib/llvm-project/libcxx/include/__exception/exception.h:113: Poco::Exception::Exception(String const&, int) @ 0x0000000022307332
2026-01-13 16:26:10 1. ./ci/tmp/build/./src/Common/Exception.cpp:136: DB::Exception::Exception(DB::Exception::MessageMasked&&, int, bool) @ 0x0000000013666da9
2026-01-13 16:26:10 2. ./src/Common/Exception.h:172: DB::Exception::Exception(String&&, int, String, bool) @ 0x000000000b1ec20e
2026-01-13 16:26:10 3. ./src/Common/Exception.h:58: DB::Exception::Exception(PreformattedMessage&&, int) @ 0x000000000b1ebcc9
2026-01-13 16:26:10 4. ./src/Common/Exception.h:190: DB::Exception::Exception<String const&>(int, FormatStringHelperImpl<std::type_identity<String const&>::type>, String const&) @ 0x000000000b1fb8f6
2026-01-13 16:26:10 5. ./ci/tmp/build/./src/TableFunctions/TableFunctionFactory.cpp:50: DB::TableFunctionFactory::get(std::shared_ptr<DB::IAST> const&, std::shared_ptr<DB::Context const>) const @ 0x000000001793d8f4
2026-01-13 16:26:10 6. ./ci/tmp/build/./src/Interpreters/Context.cpp:2595: DB::Context::executeTableFunction(std::shared_ptr<DB::IAST> const&, DB::ASTSelectQuery const*) @ 0x0000000019406403
2026-01-13 16:26:10 7. ./ci/tmp/build/./src/Interpreters/JoinedTables.cpp:221: DB::JoinedTables::getLeftTableStorage() @ 0x00000000197167f6
2026-01-13 16:26:10 8. ./ci/tmp/build/./src/Interpreters/InterpreterSelectQuery.cpp:582: DB::InterpreterSelectQuery::InterpreterSelectQuery(std::shared_ptr<DB::IAST> const&, std::shared_ptr<DB::Context> const&, std::optional<DB::Pipe>, std::shared_ptr<DB::IStorage> const&, DB::SelectQueryOptions const&, std::vector<String, std::allocator<String>> const&, std::shared_ptr<DB::StorageInMemoryMetadata const> const&, std::shared_ptr<DB::PreparedSets>) @ 0x00000000196b00d8
2026-01-13 16:26:10 9. ./ci/tmp/build/./src/Interpreters/InterpreterSelectQuery.cpp:330: DB::InterpreterSelectQuery::InterpreterSelectQuery(std::shared_ptr<DB::IAST> const&, std::shared_ptr<DB::Context> const&, DB::SelectQueryOptions const&, std::vector<String, std::allocator<String>> const&) @ 0x00000000196af291
2026-01-13 16:26:10 10. ./contrib/llvm-project/libcxx/include/__memory/unique_ptr.h:759: DB::InterpreterSelectWithUnionQuery::buildCurrentChildInterpreter(std::shared_ptr<DB::IAST> const&, std::vector<String, std::allocator<String>> const&) @ 0x00000000196eaf92
2026-01-13 16:26:10 11. ./ci/tmp/build/./src/Interpreters/InterpreterSelectWithUnionQuery.cpp:167: DB::InterpreterSelectWithUnionQuery::InterpreterSelectWithUnionQuery(std::shared_ptr<DB::IAST> const&, std::shared_ptr<DB::Context>, DB::SelectQueryOptions const&, std::vector<String, std::allocator<String>> const&) @ 0x00000000196e9c10
2026-01-13 16:26:10 12. ./ci/tmp/build/./src/Interpreters/InterpreterSelectWithUnionQuery.cpp:56: DB::InterpreterSelectWithUnionQuery::InterpreterSelectWithUnionQuery(std::shared_ptr<DB::IAST> const&, std::shared_ptr<DB::Context const>, DB::SelectQueryOptions const&, std::vector<String, std::allocator<String>> const&) @ 0x00000000196e8b42
2026-01-13 16:26:10 13. ./ci/tmp/build/./src/Interpreters/InterpreterSelectWithUnionQuery.cpp:300: DB::InterpreterSelectWithUnionQuery::getSampleBlock(std::shared_ptr<DB::IAST> const&, std::shared_ptr<DB::Context const>, bool, bool) @ 0x00000000196eb97d
2026-01-13 16:26:10 14. ./ci/tmp/build/./src/Storages/AlterCommands.cpp:872: DB::AlterCommand::apply(DB::StorageInMemoryMetadata&, std::shared_ptr<DB::Context const>) const @ 0x000000001beb6ea0
2026-01-13 16:26:10 15. ./ci/tmp/build/./src/Storages/AlterCommands.cpp:1241: DB::AlterCommands::apply(DB::StorageInMemoryMetadata&, std::shared_ptr<DB::Context const>) const @ 0x000000001bebb681
2026-01-13 16:26:10 16. ./ci/tmp/build/./src/Storages/StorageMaterializedView.cpp:673: DB::StorageMaterializedView::alter(DB::AlterCommands const&, std::shared_ptr<DB::Context const>, std::unique_lock<std::timed_mutex>&) @ 0x000000001c0288ee
2026-01-13 16:26:10 17. ./ci/tmp/build/./src/Interpreters/InterpreterAlterQuery.cpp:275: DB::InterpreterAlterQuery::executeToTable(DB::ASTAlterQuery const&) @ 0x0000000019645c4d
2026-01-13 16:26:10 18. ./ci/tmp/build/./src/Interpreters/InterpreterAlterQuery.cpp:85: DB::InterpreterAlterQuery::execute() @ 0x000000001964344d
2026-01-13 16:26:10 19. ./ci/tmp/build/./src/Interpreters/executeQuery.cpp:1738: DB::executeQueryImpl(char const*, char const*, std::shared_ptr<DB::Context>, DB::QueryFlags, DB::QueryProcessingStage::Enum, std::unique_ptr<DB::ReadBuffer, std::default_delete<DB::ReadBuffer>>&, std::shared_ptr<DB::IAST>&, std::shared_ptr<DB::ImplicitTransactionControlExecutor>, std::function<void ()>, DB::QueryResultDetails&) @ 0x0000000019a42146
2026-01-13 16:26:10 20. ./ci/tmp/build/./src/Interpreters/executeQuery.cpp:1962: DB::executeQuery(String const&, std::shared_ptr<DB::Context>, DB::QueryFlags, DB::QueryProcessingStage::Enum) @ 0x0000000019a3d7d1
2026-01-13 16:26:10 21. ./ci/tmp/build/./src/Server/TCPHandler.cpp:774: DB::TCPHandler::runImpl() @ 0x000000001ce5f108
2026-01-13 16:26:10 22. ./ci/tmp/build/./src/Server/TCPHandler.cpp:2867: DB::TCPHandler::run() @ 0x000000001ce7cce3
2026-01-13 16:26:10 23. ./ci/tmp/build/./base/poco/Net/src/TCPServerConnection.cpp:40: Poco::Net::TCPServerConnection::start() @ 0x00000000223cce47
2026-01-13 16:26:10 24. ./ci/tmp/build/./base/poco/Net/src/TCPServerDispatcher.cpp:115: Poco::Net::TCPServerDispatcher::run() @ 0x00000000223cd47e
2026-01-13 16:26:10 25. ./ci/tmp/build/./base/poco/Foundation/src/ThreadPool.cpp:205: Poco::PooledThread::run() @ 0x000000002236643f
2026-01-13 16:26:10 26. ./base/poco/Foundation/src/Thread_POSIX.cpp:341: Poco::ThreadImpl::runnableEntry(void*) @ 0x0000000022363a91
2026-01-13 16:26:10 27. ? @ 0x0000000000094ac3
2026-01-13 16:26:10 28. ? @ 0x00000000001268c0
2026-01-13 16:26:10 
2026-01-13 16:26:10 Received exception from server (version 26.1.1):
2026-01-13 16:26:10 Code: 46. DB::Exception: Received from localhost:9000. DB::Exception: Unknown table function v2:  or incorrect parameterized view. (UNKNOWN_FUNCTION)
2026-01-13 16:26:10 (query: ALTER TABLE test_79y7c1ek_1.mv1 MODIFY QUERY SELECT id, data, p FROM v2(p='modified_query_db1'))
2026-01-13 16:26:10 [7ea13ad1ee08] 2026.01.13 14:26:03.552526 [ 42956 ] {6e6c302b-cb72-4da3-a4a3-df5b3114f5c2} <Error> executeQuery: Code: 46. DB::Exception: Unknown table function v2:  or incorrect parameterized view. (UNKNOWN_FUNCTION) (version 26.1.1.468) (from [::1]:40774) (comment: 03779_materialized_view_database_context.sh-test_79y7c1ek) (query 1, line 1) (in query: ALTER TABLE test_79y7c1ek_2.mv1 MODIFY QUERY SELECT id, data, p FROM v2(p='modified_query_db2')), Stack trace (when copying this message, always include the lines below):
2026-01-13 16:26:10 
2026-01-13 16:26:10 0. ./contrib/llvm-project/libcxx/include/__exception/exception.h:113: Poco::Exception::Exception(String const&, int) @ 0x0000000022307332
2026-01-13 16:26:10 1. ./ci/tmp/build/./src/Common/Exception.cpp:136: DB::Exception::Exception(DB::Exception::MessageMasked&&, int, bool) @ 0x0000000013666da9
2026-01-13 16:26:10 2. ./src/Common/Exception.h:172: DB::Exception::Exception(String&&, int, String, bool) @ 0x000000000b1ec20e
2026-01-13 16:26:10 3. ./src/Common/Exception.h:58: DB::Exception::Exception(PreformattedMessage&&, int) @ 0x000000000b1ebcc9
2026-01-13 16:26:10 4. ./src/Common/Exception.h:190: DB::Exception::Exception<String const&>(int, FormatStringHelperImpl<std::type_identity<String const&>::type>, String const&) @ 0x000000000b1fb8f6
2026-01-13 16:26:10 5. ./ci/tmp/build/./src/TableFunctions/TableFunctionFactory.cpp:50: DB::TableFunctionFactory::get(std::shared_ptr<DB::IAST> const&, std::shared_ptr<DB::Context const>) const @ 0x000000001793d8f4
2026-01-13 16:26:10 6. ./ci/tmp/build/./src/Interpreters/Context.cpp:2595: DB::Context::executeTableFunction(std::shared_ptr<DB::IAST> const&, DB::ASTSelectQuery const*) @ 0x0000000019406403
2026-01-13 16:26:10 7. ./ci/tmp/build/./src/Interpreters/JoinedTables.cpp:221: DB::JoinedTables::getLeftTableStorage() @ 0x00000000197167f6
2026-01-13 16:26:10 8. ./ci/tmp/build/./src/Interpreters/InterpreterSelectQuery.cpp:582: DB::InterpreterSelectQuery::InterpreterSelectQuery(std::shared_ptr<DB::IAST> const&, std::shared_ptr<DB::Context> const&, std::optional<DB::Pipe>, std::shared_ptr<DB::IStorage> const&, DB::SelectQueryOptions const&, std::vector<String, std::allocator<String>> const&, std::shared_ptr<DB::StorageInMemoryMetadata const> const&, std::shared_ptr<DB::PreparedSets>) @ 0x00000000196b00d8
2026-01-13 16:26:10 9. ./ci/tmp/build/./src/Interpreters/InterpreterSelectQuery.cpp:330: DB::InterpreterSelectQuery::InterpreterSelectQuery(std::shared_ptr<DB::IAST> const&, std::shared_ptr<DB::Context> const&, DB::SelectQueryOptions const&, std::vector<String, std::allocator<String>> const&) @ 0x00000000196af291
2026-01-13 16:26:10 10. ./contrib/llvm-project/libcxx/include/__memory/unique_ptr.h:759: DB::InterpreterSelectWithUnionQuery::buildCurrentChildInterpreter(std::shared_ptr<DB::IAST> const&, std::vector<String, std::allocator<String>> const&) @ 0x00000000196eaf92
2026-01-13 16:26:10 11. ./ci/tmp/build/./src/Interpreters/InterpreterSelectWithUnionQuery.cpp:167: DB::InterpreterSelectWithUnionQuery::InterpreterSelectWithUnionQuery(std::shared_ptr<DB::IAST> const&, std::shared_ptr<DB::Context>, DB::SelectQueryOptions const&, std::vector<String, std::allocator<String>> const&) @ 0x00000000196e9c10
2026-01-13 16:26:10 12. ./ci/tmp/build/./src/Interpreters/InterpreterSelectWithUnionQuery.cpp:56: DB::InterpreterSelectWithUnionQuery::InterpreterSelectWithUnionQuery(std::shared_ptr<DB::IAST> const&, std::shared_ptr<DB::Context const>, DB::SelectQueryOptions const&, std::vector<String, std::allocator<String>> const&) @ 0x00000000196e8b42
2026-01-13 16:26:10 13. ./ci/tmp/build/./src/Interpreters/InterpreterSelectWithUnionQuery.cpp:300: DB::InterpreterSelectWithUnionQuery::getSampleBlock(std::shared_ptr<DB::IAST> const&, std::shared_ptr<DB::Context const>, bool, bool) @ 0x00000000196eb97d
2026-01-13 16:26:10 14. ./ci/tmp/build/./src/Storages/AlterCommands.cpp:872: DB::AlterCommand::apply(DB::StorageInMemoryMetadata&, std::shared_ptr<DB::Context const>) const @ 0x000000001beb6ea0
2026-01-13 16:26:10 15. ./ci/tmp/build/./src/Storages/AlterCommands.cpp:1241: DB::AlterCommands::apply(DB::StorageInMemoryMetadata&, std::shared_ptr<DB::Context const>) const @ 0x000000001bebb681
2026-01-13 16:26:10 16. ./ci/tmp/build/./src/Storages/StorageMaterializedView.cpp:673: DB::StorageMaterializedView::alter(DB::AlterCommands const&, std::shared_ptr<DB::Context const>, std::unique_lock<std::timed_mutex>&) @ 0x000000001c0288ee
2026-01-13 16:26:10 17. ./ci/tmp/build/./src/Interpreters/InterpreterAlterQuery.cpp:275: DB::InterpreterAlterQuery::executeToTable(DB::ASTAlterQuery const&) @ 0x0000000019645c4d
2026-01-13 16:26:10 18. ./ci/tmp/build/./src/Interpreters/InterpreterAlterQuery.cpp:85: DB::InterpreterAlterQuery::execute() @ 0x000000001964344d
2026-01-13 16:26:10 19. ./ci/tmp/build/./src/Interpreters/executeQuery.cpp:1738: DB::executeQueryImpl(char const*, char const*, std::shared_ptr<DB::Context>, DB::QueryFlags, DB::QueryProcessingStage::Enum, std::unique_ptr<DB::ReadBuffer, std::default_delete<DB::ReadBuffer>>&, std::shared_ptr<DB::IAST>&, std::shared_ptr<DB::ImplicitTransactionControlExecutor>, std::function<void ()>, DB::QueryResultDetails&) @ 0x0000000019a42146
2026-01-13 16:26:10 20. ./ci/tmp/build/./src/Interpreters/executeQuery.cpp:1962: DB::executeQuery(String const&, std::shared_ptr<DB::Context>, DB::QueryFlags, DB::QueryProcessingStage::Enum) @ 0x0000000019a3d7d1
2026-01-13 16:26:10 21. ./ci/tmp/build/./src/Server/TCPHandler.cpp:774: DB::TCPHandler::runImpl() @ 0x000000001ce5f108
2026-01-13 16:26:10 22. ./ci/tmp/build/./src/Server/TCPHandler.cpp:2867: DB::TCPHandler::run() @ 0x000000001ce7cce3
2026-01-13 16:26:10 23. ./ci/tmp/build/./base/poco/Net/src/TCPServerConnection.cpp:40: Poco::Net::TCPServerConnection::start() @ 0x00000000223cce47
2026-01-13 16:26:10 24. ./ci/tmp/build/./base/poco/Net/src/TCPServerDispatcher.cpp:115: Poco::Net::TCPServerDispatcher::run() @ 0x00000000223cd47e
2026-01-13 16:26:10 25. ./ci/tmp/build/./base/poco/Foundation/src/ThreadPool.cpp:205: Poco::PooledThread::run() @ 0x000000002236643f
2026-01-13 16:26:10 26. ./base/poco/Foundation/src/Thread_POSIX.cpp:341: Poco::ThreadImpl::runnableEntry(void*) @ 0x0000000022363a91
2026-01-13 16:26:10 27. ? @ 0x0000000000094ac3
2026-01-13 16:26:10 28. ? @ 0x00000000001268c0
2026-01-13 16:26:10 
2026-01-13 16:26:10 Received exception from server (version 26.1.1):
2026-01-13 16:26:10 Code: 46. DB::Exception: Received from localhost:9000. DB::Exception: Unknown table function v2:  or incorrect parameterized view. (UNKNOWN_FUNCTION)
2026-01-13 16:26:10 (query: ALTER TABLE test_79y7c1ek_2.mv1 MODIFY QUERY SELECT id, data, p FROM v2(p='modified_query_db2'))
2026-01-13 16:26:10 
2026-01-13 16:26:10 stdout:
2026-01-13 16:26:10 mv1_db1:	1	db1_data	from_mv1_db1
2026-01-13 16:26:10 mv1_db2:	2	db2_data	from_mv1_db2
2026-01-13 16:26:10 qualified:	2	db2_data	qualified_crossdb
2026-01-13 16:26:10 modified_mv1_db1:	1	db1_data	from_mv1_db1
2026-01-13 16:26:10 modified_mv1_db2:	2	db2_data	from_mv1_db2
2026-01-13 16:26:10 
2026-01-13 16:26:10 /home/ubuntu/actions-runner/_work/ClickHouse/ClickHouse/tests/queries/0_stateless/test_79y7c1ek/03779_materialized_view_database_context.sh.debuglog:
2026-01-13 16:26:10 + [2026-01-13 16:25:52] [:9] DB1=test_79y7c1ek_1
2026-01-13 16:26:10 + [2026-01-13 16:25:52] [:10] DB2=test_79y7c1ek_2
2026-01-13 16:26:10 + [2026-01-13 16:25:52] [:13] /home/ubuntu/actions-runner/_work/ClickHouse/ClickHouse/ci/tmp/clickhouse-client --send_logs_level=warning --database=test_79y7c1ek --log_comment 03779_materialized_view_database_context.sh-test_79y7c1ek --max_insert_threads 1 --group_by_two_level_threshold 584665 --group_by_two_level_threshold_bytes 50000000 --distributed_aggregation_memory_efficient 0 --fsync_metadata 1 --output_format_parallel_formatting 1 --input_format_parallel_parsing 0 --min_chunk_bytes_for_parallel_parsing 1024 --max_read_buffer_size 813162 --prefer_localhost_replica 1 --max_block_size 63354 --max_joined_block_size_rows 49555 --joined_block_split_single_row 0 --join_output_by_rowlist_perkey_rows_threshold 94 --max_threads 2 --optimize_append_index 1 --use_hedged_requests 0 --optimize_if_chain_to_multiif 1 --optimize_if_transform_strings_to_enum 0 --optimize_read_in_order 1 --optimize_or_like_chain 0 --optimize_substitute_columns 0 --enable_multiple_prewhere_read_steps 1 --read_in_order_two_level_merge_threshold 15 --optimize_aggregation_in_order 1 --aggregation_in_order_max_block_bytes 9674430 --use_uncompressed_cache 1 --min_bytes_to_use_direct_io 10737418240 --min_bytes_to_use_mmap_io 588074974 --local_filesystem_read_method pread_threadpool --remote_filesystem_read_method threadpool --local_filesystem_read_prefetch 1 --filesystem_cache_segments_batch_size 3 --read_from_filesystem_cache_if_exists_otherwise_bypass_cache 1 --throw_on_error_from_cache_on_write_operations 0 --remote_filesystem_read_prefetch 1 --allow_prefetched_read_pool_for_remote_filesystem 0 --filesystem_prefetch_max_memory_usage 32Mi --filesystem_prefetches_limit 10 --filesystem_prefetch_min_bytes_for_single_read_task 1Mi --filesystem_prefetch_step_marks 0 --filesystem_prefetch_step_bytes 0 --compile_expressions 1 --compile_aggregate_expressions 1 --compile_sort_description 0 --merge_tree_coarse_index_granularity 15 --optimize_distinct_in_order 0 --max_bytes_before_remerge_sort 2006539996 --min_compress_block_size 1465431 --max_compress_block_size 1966313 --merge_tree_compact_parts_min_granules_to_multibuffer_read 124 --optimize_sorting_by_input_stream_proper
FAIL	Scrapping system tables [cidb](https://play.clickhouse.com/play?user=play&run=1#V0lUSAogICAgOTAgQVMgaW50ZXJ2YWxfZGF5cwpTRUxFQ1QKICAgIHRvU3RhcnRPZkRheShjaGVja19zdGFydF90aW1lKSBBUyBkYXksCiAgICBjb3VudCgpIEFTIGZhaWx1cmVzLAogICAgZ3JvdXBVbmlxQXJyYXkocHVsbF9yZXF1ZXN0X251bWJlcikgQVMgcHJzLAogICAgYW55KHJlcG9ydF91cmwpIEFTIHJlcG9ydF91cmwKRlJPTSBjaGVja3MKV0hFUkUgKG5vdygpIC0gdG9JbnRlcnZhbERheShpbnRlcnZhbF9kYXlzKSkgPD0gY2hlY2tfc3RhcnRfdGltZQogICAgQU5EIHRlc3RfbmFtZSA9ICdTY3JhcHBpbmcgc3lzdGVtIHRhYmxlcycKICAgIC0tIEFORCBjaGVja19uYW1lID0gJ1N0YXRlbGVzcyB0ZXN0cyAoYW1kX2JpbmFyeSwgb2xkIGFuYWx5emVyLCBzMyBzdG9yYWdlLCBEYXRhYmFzZVJlcGxpY2F0ZWQsIHBhcmFsbGVsKScKICAgIEFORCB0ZXN0X3N0YXR1cyBJTiAoJ0ZBSUwnLCAnRVJST1InKQogICAgQU5EIChwdWxsX3JlcXVlc3RfbnVtYmVyID0gMCBPUiBiYXNlX3JlZiBJTiAoJ21hc3RlcicpKQogICAgQU5EIHRlc3RfY29udGV4dF9yYXcgTElLRSAnJUNvZGU6JScKR1JPVVAgQlkgZGF5Ck9SREVSIEJZIGRheSBERVNDCg==)[issue](https://github.com/ClickHouse/ClickHouse/issues/93959)
```

dk-github committed:
Merge branch 'ClickHouse:master' into issue-87082-fix-materialized-views-name-context

dk-github committed:
fixes ClickHouse#87082 - materialized views to use database where they were created as context - addressing review comments - one more place to set correct context, plus extended and clarified test for all different cases

dk-github commented on tests/queries/0_stateless/03779_materialized_view_database_context.sh:
```diff
@@ -0,0 +1,61 @@
+$CLICKHOUSE_CLIENT --query "CREATE TABLE $DB1.tbl1 (id Int32, data String) ENGINE = MergeTree ORDER BY id"
+$CLICKHOUSE_CLIENT --query "INSERT INTO $DB1.tbl1 VALUES (1, 'db1_data')"
+$CLICKHOUSE_CLIENT --query "CREATE VIEW $DB1.v1 AS SELECT id, data, {p:String} as p FROM $DB1.tbl1"
+$CLICKHOUSE_CLIENT --query "CREATE MATERIALIZED VIEW $DB1.mv1 REFRESH EVERY 1 SECOND ORDER BY id AS SELECT id, data, p FROM $DB1.v1(p='from_mv1_db1')"
```
Got it, thank you! yes, this is yet another place where the same fix was required - passing correct context to query analysis. (I must say, I didn't realize when I started it that there would be so many disparate places where the same internal sql is going to be analysed...) Hopefully this is the last place - thank you for spotting it!

So I added the fix for this case, and also modified the test to more clearly show what I originally fixed, then added your test practically verbatim at the end - it also passes now. the only thing I had to change was making sue that "use database $db2" is executed in one client invocation with the rest of it - but it is not super-important, it should work as you wrote it as well - with default session database

please have another look when you have time

dk-github committed:
fixes ClickHouse#87082 - materialized views to use database where they were created as context - discovered that there apparently is another, "old" query analyser, and the same context-related changes also need to be made in a bunch of "old" code paths...

dk-github committed:
Merge branch 'ClickHouse:master' into issue-87082-fix-materialized-views-name-context

dk-github commented:
> Some tests failed: https://s3.amazonaws.com/clickhouse-test-reports/json.html?PR=88193&sha=33b6fdfdb95d104d4e1514715b11084a21b85e49&name_0=PR

I was confused because the first set of tests in teh link were clearly failing due to some (intermittent?) ZK issues, but looking at the second revealed that while my test passed in other configs, the "old analyser" config fails, and there is actually a whole bunch of "old analyser" codepaths which I had never encountered that needed a similar treatment to the one in "new" analyzer... thanks for spotting, it passes now when i run it locally with "old analyser" settings too...

pamarcos committed:
Refactor for consistency

Put SQL statements together when they fit together to avoid
making too many client calls. This improves the test 1.5x.

pamarcos committed:
Merge branch 'master' into issue-87082-fix-materialized-views-name-context

dk-github commented:
> There are some tests that fail quite consistently with runs for this PR:

okay, let me see if I can reproduce, or find anything  in the various logs that is more relevant than the exception bubbled up to the test runner... I see in the last update that only 3 are failing (below) - so will start with those ones


02232_allow_only_replicated_engine 
02514_database_replicated_no_arguments_for_rmt 
03248_create-same-table-concurrently-with-replicated-engine

dk-github committed:
Merge branch 'ClickHouse:master' into issue-87082-fix-materialized-views-name-context

dk-github committed:
Merge branch 'ClickHouse:master' into issue-87082-fix-materialized-views-name-context

dk-github commented:
@pamarcos I have merged in a bunch of latest changes from main, which included fixes for various flakiness modes, and the latest run failed in just ONE test, which is:
- not one of the previous often-failing ones which you were worried about (those passed)
- not from the same build/architecture suite as last time, so any link to specific architecture seems even less likely
- doesn't seem to have anything to do with my changes
- has retry_ok label - not sure what that means, if not that it passed on retry?

I think this reinforces my argument that my changes are passing all tests and look good to go - but of course if you have any concerns left, please let me know what should my next steps be to get this over the line.

Thanks for your time!

pamarcos committed:
Fix for issue found in private tests

pamarcos reviewed (approved):
Thanks @dk-github for your contribution. Much appreciated.
Sorry it took a while, but the current CI status is so-so and there was also the private CI that was failing due to your changes but it was impossible for you to identify or fix since you don't have access to it.
